torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp

```diff
@@ -94,116 +25,80 @@ void initNvFuserPythonBindings(PyObject* module) {
-                "TensorView does not have DataType?");
-            ss << self.getDataType().value();
-            return self.toString() + " DataType: " + ss.str() +
-                " Contiguity: " + self.domain()->getContiguityString();
-          },
-          py::return_value_policy::reference);
-  py::class_<torch::jit::fuser::cuda::Val>(nvfuser, "Val")
-      .def(
-          "__str__",
-          [](torch::jit::fuser::cuda::Val& self) -> std::string {
-            return self.toString();
-          },
-          py::return_value_policy::reference);
+  py::class_<nvfuser::Tensor>(nvfuser, "Tensor");
+  py::class_<nvfuser::Scalar>(nvfuser, "Scalar");
 
   // C++ Side of Context Manager used to mimic the FusionGuard as a way
   // to programatically distinguish code used to define the Fusion instead
   // of having the user mysteriously create an object prior to adding definition
   // code where the object is not used.
-  py::class_<FusionDefinitionContextManager> fusion_def(
-      nvfuser, "FusionDefinition");
-  fusion_def.def(py::init<PythonFusionOwner*>())
+  py::class_<nvfuser::FusionDefinition> fusion_def(nvfuser, "FusionDefinition");
+  fusion_def.def(py::init<nvfuser::FusionOwner*>())
+      .def_readwrite("ops", &nvfuser::FusionDefinition::ops)
       .def(
           "__enter__",
-          [](FusionDefinitionContextManager& self) { return self.enter(); })
+          [](nvfuser::FusionDefinition& self)->nvfuser::FusionDefinition* { return self.enter(); })
       .def(
           "__exit__",
-          [](FusionDefinitionContextManager& self,
+          [](nvfuser::FusionDefinition& self,
              void* exc_type,
              void* exc_value,
-             void* traceback) { self.exit(); })
-      .def(
-          "add_input",
-          [](FusionDefinitionContextManager& self,
-             torch::jit::fuser::cuda::Val* input) { self.addInput(input); })
-      .def(
-          "add_input",
-          [](FusionDefinitionContextManager& self, TensorView* input) {
-            self.addInput(input);
+             void* traceback) { 
+               self.fusion_state.resize(self.recording_state.size(), nullptr);
+               for (auto& record : self.recording) {
+                 auto functor = record.get();
+                 (*functor)(self);
+               }
+               self.exit();
           })
       .def(
           "add_output",
-          [](FusionDefinitionContextManager& self,
-             torch::jit::fuser::cuda::Val* output) { self.addOutput(output); })
+          [](nvfuser::FusionDefinition& self, nvfuser::Scalar* output) { 
+            self.recording.emplace_back(
+                new nvfuser::OutputRecord<NvfVal>({output->index}));
+          })
       .def(
           "add_output",
-          [](FusionDefinitionContextManager& self, TensorView* output) {
-            self.addOutput(output);
+          [](nvfuser::FusionDefinition& self, nvfuser::Tensor* output) {
```

Comment:
We might want to revisit the signature here. But that can be in a different topic.